romtools/workflows/sampling_with_holdout/sampling_with_holdout.py

```diff
@@ -0,0 +1,227 @@
+from romtools.workflows.sampling import *
+from romtools.workflows.model_builders import QoiModelBuilder
+
+
+def _create_parameter_dict(parameter_names, parameter_values):
+    return dict(zip(parameter_names, parameter_values))
+
+
+def run_sampling_with_holdout(
+    fom_model: QoiModel,
+    rom_model_builder: QoiModelBuilder,
+    parameter_space: ParameterSpace,
+    absolute_work_directory: str,
+    holdout_set_size: int = 5,
+    max_number_of_rom_samples: int = 20,
+    tolerance=1e-5,
+    random_seed: int = 1,
+):
+    """
+    Core algorithm
+    """
+    sampling_directory = absolute_work_directory
+    create_empty_dir(sampling_directory)
+    offline_directory_prefix = "offline_data"
+
+    run_directory_prefix = "run_"
+    sampling_file = open(
+        f"{sampling_directory}/sampling_with_holdout_status.log", "w", encoding="utf-8"
+    )
+    sampling_file.write("Holdout sampling status \n")
+    fom_time = 0.0
+    rom_time = 0.0
+    basis_time = 0.0
+
+    np.random.seed(random_seed)
+
+    parameter_samples = parameter_space.generate_samples(
+        holdout_set_size + max_number_of_rom_samples
+    )
+    parameter_names = parameter_space.get_names()
+
+    holdout_sample_indices = range(holdout_set_size)
+    training_sample_indices = range(
+        holdout_set_size, holdout_set_size + max_number_of_rom_samples
+    )
+    holdout_samples = parameter_samples[holdout_sample_indices]
+    training_samples = parameter_samples[training_sample_indices]
+
+    # Setup FOM directories for holdout set and potential training runs
+    for sample_index, sample in enumerate(holdout_samples):
+        fom_run_directory = (
+            f"{sampling_directory}/fom/holdout_set/{run_directory_prefix}{sample_index}"
+        )
+        create_empty_dir(fom_run_directory)
+
+        parameter_dict = _create_parameter_dict(parameter_names, sample)
+        fom_model.populate_run_directory(fom_run_directory, parameter_dict)
+
+    for sample_index, sample in enumerate(training_samples):
+        fom_run_directory = f"{sampling_directory}/fom/training_set/{run_directory_prefix}{sample_index}"
+        create_empty_dir(fom_run_directory)
+
+        parameter_dict = _create_parameter_dict(parameter_names, sample)
+        fom_model.populate_run_directory(fom_run_directory, parameter_dict)
+
+    # Run FOM samples to build holdout set.
+    t0 = time.time()
+    fom_qois_holdout_set = np.zeros(holdout_set_size)
+    sampling_file.write(f"Building holdout set \n")
+    for i in holdout_sample_indices:
+        sampling_file.write(f"Running holdout FOM sample {i} \n")
+        sample_index = i
+        parameter_dict = _create_parameter_dict(
+            parameter_names, parameter_samples[sample_index]
+        )
+        fom_run_directory = (
+            f"{sampling_directory}/fom/holdout_set/{run_directory_prefix}{sample_index}"
+        )
+        create_empty_dir(fom_run_directory)
```

Comment:
Wasn't this directory already created on line 106?

Reply:
True. Removed lines 106-109 in the latest commit. Can also just remove these 3 lines if preferred.